source/MaterialXGenShader/Nodes/CombineNode.cpp

```diff
@@ -0,0 +1,127 @@
+#include <MaterialXGenShader/Nodes/CombineNode.h>
+#include <MaterialXGenShader/ShaderGenerator.h>
+#include <MaterialXGenShader/TypeDesc.h>
+
+namespace MaterialX
+{
+
+ShaderNodeImplPtr CombineNode::create()
+{
+    return std::make_shared<CombineNode>();
+}
+
+void CombineNode::emitFunctionCall(const ShaderNode& node, GenContext& context, ShaderGenerator& shadergen, Shader& shader)
+{
+    BEGIN_SHADER_STAGE(shader, Shader::PIXEL_STAGE)
+
+    const ShaderInput* in1 = node.getInput(0);
+    const ShaderOutput* out = node.getOutput();
+    if (!in1 || !out)
+    {
+        throw ExceptionShaderGenError("Node '" + node.getName() + "' is not a valid convert node");
+    }
+
+    // Check the node signature to determine which
+    // type conversion to perform, and get the value
+    // components to use for constructing the new value.
+    //
+    vector<string> valueComponents;
+    if (in1->type == Type::FLOAT)
+    {
+        // Get the components of the input values.
+        const size_t numInputs = node.numInputs();
+        valueComponents.resize(numInputs);
+        for (size_t i = 0; i < numInputs; ++i)
+        {
+            const ShaderInput* input = node.getInput(i);
+            shadergen.getInput(context, input, valueComponents[i]);
+        }
+
+    }
+    else if (in1->type == Type::COLOR3 || in1->type == Type::VECTOR3)
+    {
+        const ShaderInput* in2 = node.getInput(1);
+        if (!in2 || in2->type != Type::FLOAT)
+        {
+            throw ExceptionShaderGenError("Node '" + node.getName() + "' is not a valid convert node");
+        }
+
+        // If in1 is unconnected we must declare a local variable
+        // for it first, in order to access components from it below.
+        string in1Variable = in1->connection ? in1->connection->variable : in1->variable;
+        if (!in1->connection)
+        {
+            string variableValue = in1->value ? shadergen.getSyntax()->getValue(in1->type, *in1->value) : shadergen.getSyntax()->getDefaultValue(in1->type);
+            shader.addLine(shadergen.getSyntax()->getTypeName(in1->type) + " " + in1Variable + " = " + variableValue);
+        }
```

Comment:
This is the main reason why a C++ class is needed. A local variable needs to be used if the input is unconnected. Otherwise we will reference a component directly on the value constructor, and this is not supported in all languages. For example this is invalid in OSL: `color(1.0, 1.0, 1.0)[2]`